File: keeping up with the javascripts part 1 es6/Homeworks/13 - Classes/vehicles.js
/*
 * Homework #13: Classes
 * Keeping Up With the Javascripts - Part 1: ES6
 * Pirple
 * 
 * Alejandro Mohamad
 */

class Vehicle {
  constructor(make, model, year, weight) {
    this.make = make;
    this.model = model;
    this.year = year;
    this.weight = weight;
    this.needsMaintenance = false;
    this.tripsSinceMaintenance = 0;
  }

  resetMaintenance() {
    this.needsMaintenance = false;
    this.tripsSinceMaintenance = 0;
  }

  repair() {
    this.resetMaintenance();
  }

  displayInformation() {
    console.log(`Make: ${this.make}`);
    console.log(`Model: ${this.model}`);
    console.log(`Year: ${this.year}`);
    console.log(`Weight: ${this.weight}`);
    console.log(`- Needs maintenance: ${this.needsMaintenance}`);
    console.log(`- Trips since maintenance: ${this.tripsSinceMaintenance}`);
    console.log(`------`);
  }
}

class Cars extends Vehicle {
  drive() {
    this.isDriving = true;
  }
  stop() {
    this.isDriving = false;
    this.tripsSinceMaintenance += 1;

    if (this.tripsSinceMaintenance === 100) {
      this.needsMaintenance = true;
    }
  }
}

const tesla = new Cars('Tesla', 'Model X', 2015, 2300);
const subaru = new Cars('Subaru', 'XV', 2011, 2200);
const citroen = new Cars('Citroën', 'C3', 2001, 1005);

for (let i = 0; i < 10; i++) {
  tesla.drive();
  tesla.stop();
}

for (let i = 0; i < 58; i++) {
  subaru.drive();
  subaru.stop();
}

for (let i = 0; i < 112; i++) {
  citroen.drive();
  citroen.stop();
}

tesla.displayInformation();
subaru.displayInformation();
citroen.displayInformation();

citroen.repair();
citroen.displayInformation();

/*
 * Extra Credit
 */

class Planes extends Vehicle {
  flying() {
    if (this.needsMaintenance) {
      console.log(`The plane ${this.make} ${this.model} can't fly until it's repaired`);
      return false;
    }

    this.isFlying = true;
  }
  landing() {
    if (this.isFlying === false) {
      return false;
    }

    this.isFlying = false;
    this.tripsSinceMaintenance += 1;

    if (this.tripsSinceMaintenance === 100) {
      this.needsMaintenance = true;
    }
  }
}

const airbus = new Planes('Airbus', 'A350', 2013, 45000);
const boeing = new Planes('Boeing', '787', 2009, 60000);

for (let i = 0; i < 70; i++) {
  airbus.flying();
  airbus.landing();
}

for (let i = 0; i < 103; i++) {
  boeing.flying();
  boeing.landing();
}

airbus.displayInformation();
boeing.displayInformation();

boeing.flying();

File: get swifty part 1/Homeworks/05 - Classes, Structs, and Enums/vehicles.swift
/*
 *  Homework #5: Classes, Structs, and Enums
 *  Get Swifty - Part 1
 *  Pirple
 *
 *  Alejandro Mohamad
 */

class Vehicle {
    var make: String
    var model: String
    var year: Int
    var weight: Int
    
    var needsMaintenance = false
    var tripsSinceMaintenance = 0
    
    init(make: String, model: String, year: Int, weight: Int) {
        self.make = make
        self.model = model
        self.year = year
        self.weight = weight
    }
    
    func repair() {
        needsMaintenance = false
        tripsSinceMaintenance = 0
    }
    
    func displayInformation() {
        print("Make: \(make)")
        print("Model: \(model)")
        print("Year: \(year)")
        print("Weight: \(weight)")
        print("- Needs maintenance: \(needsMaintenance)")
        print("- Trips since maintenance: \(tripsSinceMaintenance)")
        print("------")
    }
}

class Cars: Vehicle {
    var isDriving = false
    
    func drive() {
        isDriving = true
    }
    
    func stop() {
        isDriving = false
        tripsSinceMaintenance += 1
        
        if tripsSinceMaintenance == 100 {
            needsMaintenance = true
        }
    }
}

let tesla = Cars(make: "Tesla", model: "Model X", year: 2015, weight: 2300)
let subaru = Cars(make: "Subaru", model: "XV", year: 2011, weight: 2200)
let citroen = Cars(make: "Citroën", model: "C3", year: 2001, weight: 1005)

for _ in 1..<10 {
    tesla.drive()
    tesla.stop()
}

for _ in 1..<58 {
    subaru.drive()
    subaru.stop()
}

for _ in 1..<112 {
    citroen.drive()
    citroen.stop()
}

tesla.displayInformation()
subaru.displayInformation()
citroen.displayInformation()

citroen.repair()
citroen.displayInformation()

/*
 * Extra Credit
 */

class Planes: Vehicle {
    var isFlying = false
    
    func flying() {
        if needsMaintenance {
            print("The plane \(make) \(model) can't fly until it's repaired")
            return
        }
        
        isFlying = true
    }
    
    func landing() {
        if isFlying == false {
            return
        }
        
        isFlying = false
        tripsSinceMaintenance += 1
        
        if tripsSinceMaintenance == 100 {
            needsMaintenance = true
        }
    }
}

let airbus = Planes(make: "Airbus", model: "A350", year: 2013, weight: 45000)
let boeing = Planes(make: "Boeing", model: "787", year: 2009, weight: 60000)

for _ in 1..<70 {
    airbus.flying()
    airbus.landing()
}

for _ in 1..<103 {
    boeing.flying()
    boeing.landing()
}

airbus.displayInformation()
boeing.displayInformation()

boeing.flying()
